presto-delta/src/main/java/com/facebook/presto/delta/DeltaClient.java

```diff
@@ -70,72 +77,101 @@ public DeltaClient(HdfsEnvironment hdfsEnvironment)
      */
     public Optional<DeltaTable> getTable(
+            DeltaConfig config,
             ConnectorSession session,
             SchemaTableName schemaTableName,
             String tableLocation,
             Optional<Long> snapshotId,
             Optional<Long> snapshotAsOfTimestampMillis)
     {
-        Optional<DeltaLog> deltaLog = loadDeltaTableLog(session, new Path(tableLocation), schemaTableName);
-        if (!deltaLog.isPresent()) {
+        Path location = new Path(tableLocation);
+        Optional<Engine> deltaEngine = loadDeltaTableClient(session, location, schemaTableName);
+        if (!deltaEngine.isPresent()) {
             return Optional.empty();
         }
 
+        Table deltaTable = loadDeltaTable(location.toString(), deltaEngine.get());
+        Snapshot snapshot = getSnapshot(deltaTable, deltaEngine.get(), schemaTableName, snapshotId,
+                snapshotAsOfTimestampMillis);
+        return Optional.of(new DeltaTable(
+                schemaTableName.getSchemaName(),
+                schemaTableName.getTableName(),
+                tableLocation,
+                Optional.of(snapshot.getVersion(deltaEngine.get())), // lock the snapshot version
+                getSchema(config, schemaTableName, deltaEngine.get(), snapshot)));
+    }
+
+    private Snapshot getSnapshot(
+            Table deltaTable,
+            Engine deltaEngine,
+            SchemaTableName schemaTableName,
+            Optional<Long> snapshotId,
+            Optional<Long> snapshotAsOfTimestampMillis)
+    {
         // Fetch the snapshot info for given snapshot version. If no snapshot version is given, get the latest snapshot info.
         // Lock the snapshot version here and use it later in the rest of the query (such as fetching file list etc.).
         // If we don't lock the snapshot version here, the query may end up with schema from one version and data files from another
         // version when the underlying delta table is changing while the query is running.
         Snapshot snapshot;
         if (snapshotId.isPresent()) {
-            snapshot = getSnapshotById(deltaLog.get(), snapshotId.get(), schemaTableName);
+            snapshot = getSnapshotById(deltaTable, deltaEngine, snapshotId.get(), schemaTableName);
         }
         else if (snapshotAsOfTimestampMillis.isPresent()) {
-            snapshot = getSnapshotAsOfTimestamp(deltaLog.get(), snapshotAsOfTimestampMillis.get(), schemaTableName);
+            snapshot = getSnapshotAsOfTimestamp(deltaTable, deltaEngine,
+                    snapshotAsOfTimestampMillis.get(), schemaTableName);
         }
         else {
-            snapshot = deltaLog.get().snapshot(); // get the latest snapshot
+            try {
+                snapshot = deltaTable.getLatestSnapshot(deltaEngine); // get the latest snapshot
+            }
+            catch (TableNotFoundException e) {
+                throw new PrestoException(GENERIC_INTERNAL_ERROR,
```

Comment:
embed the original exception as the cause here

Reply:
Done. Added new DeltaErrorCodes to candle these cases